src/aws-cpp-sdk-core/include/smithy/interceptor/impl/ChecksumInterceptor.h

```diff
@@ -0,0 +1,209 @@
+/**
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+#include <smithy/interceptor/Interceptor.h>
+
+#include <aws/core/utils/crypto/PrecalculatedHash.h>
+#include <aws/core/utils/crypto/CRC32.h>
+#include <aws/core/utils/crypto/Sha1.h>
+#include <aws/core/utils/crypto/Sha256.h>
+#include <aws/core/utils/HashingUtils.h>
+
+namespace smithy
+{
+    namespace interceptor
+    {
+        static const char* CHECKSUM_INTERCEPTOR_LOG_TAG = "ChecksumInterceptor";
+
+        class ChecksumInterceptor final : public Interceptor
+        {
+        public:
+
+            ChecksumInterceptor() = default;
+            ChecksumInterceptor(const ChecksumInterceptor& other) = delete;
+            ChecksumInterceptor(ChecksumInterceptor&& other) noexcept = default;
+            ChecksumInterceptor& operator=(const ChecksumInterceptor& other) = delete;
+            ChecksumInterceptor& operator=(ChecksumInterceptor&& other) noexcept = default;
+            ~ChecksumInterceptor() override = default;
+
+            ModifyRequestOutcome ModifyRequest(InterceptorContext& context) override
```

Comment:
all of the code in this file is copied from [the existing implentation](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/client/AWSClient.cpp#L799). I will refactor the logic in the next PR. This PR more or less jsut refactors the code out to a single location before handling the actual logical refactor. So while we can address comments on the logic here, a lot will be changing to it in the next PR.